### ISSUE TITLE
An API question - guard based on event parameter

### ISSUE THREAD

**Issue by sorokod** [nsk90/kstatemachine#69]
I have a `MyDataEvent` that is a `DataEvent<Int>`. A transition should occur only if data is non negative (for example). 

It looks like a `guard` thing but I don't see a way of accessing the event data from a guard. 

```kotlin
transition<MyDataEvent> {
    guard = { event.data >= 0 } // ????
    targetState = SomeState
}
```

Am I missing something? Is there a better approach?

**Comment by nsk90** [comment]
Hello!
Check out this sample https://github.com/nsk90/kstatemachine/blob/master/samples/src/commonMain/kotlin/ru/nsk/samples/TypesafeTransitionSample.kt

It looks very close to your code, but if you use `DataState`/`DataEvent` api, I assume that you should use `dataTransition` instead of simple `transition` function.

**Comment by sorokod** [comment]
Ok, interestingly in this case the `targetState` must be a `DataState`. There may be use cases where this is not necessary.

**Comment by nsk90** [comment]
Yes, `DataEvent` is designed to be used in a couple with `DataState`. 
That is the way how I can guarantee at compile time that state receives expected data. If you don't need compile time guarantee or this API does not completely fit you case, you can subclass simple `Event`, not `DataEvent`.

like this:
```
MyEvent(val data) : Event
```
This block should work in this case `guard = { event.data >= 0 }`

**Comment by sorokod** [comment]
Nice, that is what I was looking for.

**Comment by nsk90** [comment]
Hello
